pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -32,61 +34,38 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
 		if err != nil {
 			return nil, err
 		}
-		version, hasVersion = int(v), true
+		version, hasVersion = v, true
 	}
 
-	// First, build a JSON state from the InstanceState.
-	json, version, err := schema.UpgradeFlatmapState(ctx, version, m, res, p.Meta())
-	if err != nil {
-		return nil, err
+	var (
+		rawState cty.Value
+		err      error
+	)
+	switch GetInstanceStateStrategy(v2Resource{res}) {
+	// Opt-out of normalization under a flag. Normalization replaces nulls with empty blocks for the state value,
+	// but when processing config the codebase calls NewResourceConfigShimmed which in turn calls
+	// ConfigValueFromHCL2Block which removes empty blocks. This leads to non-empty diffs.
+	case CtyInstanceState:
+		rawState, err = upgradeResourceStateNonNormalizing(ctx, m, p, int(version), res)
+	default:
+		rawState, err = upgradeResourceStateRPC(ctx, typeName, m, p, res, version)
```

Comment:
This is a very tedious spot-the-differences exercise hiding here. I've read the before/after code line by line and they seem to be following very closely. 

Before:

```go
json, _, _ := UpgradeFlatmapState(_, _, m, _, _)
json, _ = UpgradeJSONState(json)
RemoveAttributes(_, json, _)
v, _ = JSONMapToStateValue(json, configBlock)
v, _ = CoerceValue(v)
v = schema.NormalizeObjectFromLegacySDK(v, configBlock)
```

After:

```go
func (s *GRPCProviderServer) UpgradeResourceState(
   s.upgradeFlatmapState(ctx, version, req.RawState.Flatmap, res)
   jsonMap, err = s.upgradeJSONState(ctx, version, jsonMap, res)

   RemoveAttributes(ctx, jsonMap, schemaBlock.ImpliedType())
   val, err := JSONMapToStateValue(jsonMap, schemaBlock)
   val, err = schemaBlock.CoerceValue(val)
   val = objchange.NormalizeObjectFromLegacySDK(val, schemaBlock)
   // serialize to msgpack

// deserialize from msgpack        
var newState cty.Value

newState, err := res.ShimInstanceStateFromValue(rawState)
```

It seems that the serialization of cty.Value to msgpack and back is the only addition. That one I think should not be lossy. 

It seems like the previous method was an inlining of the logic in the gRPC provider.

Reply:
That is my understanding, which I why this is a no-op refactor, not a bug fix.